test/perf/hashtable/hashtable.cc

```diff
@@ -0,0 +1,282 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#include <cpp/when.h>
+#include <debug/harness.h>
+#include <memory>
+#include <set>
+
+/**
+ * This benchmarks test different performance of reader-writer cowns.
+ * It creates a hashtable with N buckets (num_buckets). Each behaviour can
+ * acquire M buckets (num_dependent_buckets) in read or write mode (rw_ratio /
+ * rw_ratio_denom). A behaviour can also spin for extra time to simulate
+ * short/long read (read_loop_count) and write (write_loop_count) critical
+ * sections. The benchmark creates X behaviours (num_operations) before
+ * executing them.
+ */
+
+/**
+ * This knob controls whether reader-writer properties are checked during
+ * behaviour execution. The check ensures that only readers (one or more) or a
+ * single writer is running at a given point in time.
+ */
+
+#define DEBUG_RW 0
+
+using namespace verona::cpp;
+
+long num_buckets = 16;
+long num_dependent_buckets = 8;
+long num_entries_per_bucket = 32;
+long num_operations = 10'000;
+long rw_ratio = 90;
+long rw_ratio_denom = 100;
+long read_loop_count = 0;
+long write_loop_count = 0;
+
+class Entry
+{
+public:
+  union
+  {
+    size_t val;
+    char alignment[64];
+  };
+  Entry(size_t val) : val(val) {}
+};
+
+class Bucket
+{
+public:
+  std::vector<std::shared_ptr<Entry>> list;
+
+  Bucket(std::vector<std::shared_ptr<Entry>> list1) {}
+
+  uint64_t get_addr()
+  {
+    return (uint64_t)this;
+  }
+};
+
+thread_local long found_read_ops = 0;
+thread_local long not_found_read_ops = 0;
+thread_local long found_write_ops = 0;
+thread_local long not_found_write_ops = 0;
+thread_local long mixed_ops = 0;
+thread_local long read_cs_time = 0;
+thread_local long write_cs_time = 0;
+
+std::atomic<long> total_found_read_ops = 0;
+std::atomic<long> total_not_found_read_ops = 0;
+std::atomic<long> total_found_write_ops = 0;
+std::atomic<long> total_not_found_write_ops = 0;
+std::atomic<long> total_mixed_ops = 0;
+std::atomic<long> total_read_cs_time = 0;
+std::atomic<long> total_write_cs_time = 0;
+
+#if DEBUG_RW
+auto concurrency = std::make_shared<std::array<std::atomic<size_t>, 1024>>();
+#endif
+
+void test_hash_table()
+{
+  auto t1 = high_resolution_clock::now();
+
```

Comment:
Use our PRNG, this can be seeded based on systematic testing, and will be cross platform debuggable.  The standard library is free to define its randomness for a platform, which makes cross platform systematic testing hard.
```suggestion
    xoroshiro::p128r32 rand{Systematic::get_prng_next()}; 

```